File: client-sdk/godot/AethelgardCoopClient.ts
/*
 * AETHELGRAD Godot TypeScript networking wrapper.
 *
 * Compatible with Godot TypeScript runtimes that expose the standard fetch API.
 * Keep access and refresh tokens in platform-secure storage in production.
 */

export type AuthBundle = {
  accessToken: string;
  refreshToken: string;
  tokenType: "Bearer";
  accountId: string;
  accountType?: "guest" | "google";
  expiresAt: string;
  refreshExpiresAt: string;
};

export type Participant = {
  accountId: string;
  playerX: number;
  playerY: number;
  atTower: boolean;
  towerRevision: number;
};

export type RoomSnapshot = {
  code: string;
  region: string;
  maxPlayers: number;
  worldTime: number;
  towerRevision: number;
  bossHealth: number;
  combatRevision: number;
  participants: Participant[];
};

export type RoomResponse = {
  room: RoomSnapshot;
  participants: Participant[];
};

export type CombatResult = {
  accepted: boolean;
  action: "attack" | "heavy_attack";
  targetId: string;
  damage: number;
  bossHealth: number;
  combatRevision: number;
};

export type InventoryState = {
  wood: number;
  fiber: number;
  stone: number;
  emberKit: boolean;
};

export type InventoryResult = {
  accepted: boolean;
  operation: "gather" | "craft";
  inventory: InventoryState;
  inventoryRevision: number;
};

export type HeartbeatInput = {
  playerX: number;
  playerY: number;
  atTower: boolean;
  towerRevision: number;
};

export type ClientOptions = {
  baseUrl: string;
  accessToken?: string;
  refreshToken?: string;
  fetchImpl?: typeof fetch;
  onAuthBundle?: (bundle: AuthBundle) => void;
  onRoomSnapshot?: (response: RoomResponse) => void;
};

export class AethelgardApiError extends Error {
  readonly status: number;
  readonly code: string;

  constructor(status: number, code: string) {
    super(`AETHELGRAD API ${status}: ${code}`);
    this.name = "AethelgardApiError";
    this.status = status;
    this.code = code;
  }
}

export class AethelgardCoopClient {
  private readonly baseUrl: string;
  private readonly fetchImpl: typeof fetch;
  private accessToken: string | null;
  private refreshToken: string | null;
  private requestCounter = 0;
  private readonly onAuthBundle?: (bundle: AuthBundle) => void;
  private readonly onRoomSnapshot?: (response: RoomResponse) => void;

  constructor(options: ClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, "");
    this.fetchImpl = options.fetchImpl ?? fetch;
    this.accessToken = options.accessToken ?? null;
    this.refreshToken = options.refreshToken ?? null;
    this.onAuthBundle = options.onAuthBundle;
    this.onRoomSnapshot = options.onRoomSnapshot;
  }

  setTokens(bundle: Pick<AuthBundle, "accessToken" | "refreshToken">): void {
    this.accessToken = bundle.accessToken;
    this.refreshToken = bundle.refreshToken;
  }

  clearTokens(): void {
    this.accessToken = null;
    this.refreshToken = null;
  }

  async exchangeGoogleIdToken(idToken: string): Promise<AuthBundle> {
    const bundle = await this.request<AuthBundle>("POST", "/v1/auth/google-id-token/exchange", { idToken }, false);
    this.setTokens(bundle);
    this.onAuthBundle?.(bundle);
    return bundle;
  }

  async authenticateGuest(guestKey: string): Promise<AuthBundle> {
    if (!/^[A-Za-z0-9_-]{32,128}$/.test(guestKey)) throw new Error("guestKey must be a base64url secret with 32-128 characters");
    const bundle = await this.request<AuthBundle>("POST", "/v1/auth/guest", { guestKey }, false);
    this.setTokens(bundle);
    this.onAuthBundle?.(bundle);
    return bundle;
  }

  async refresh(): Promise<AuthBundle> {
    if (!this.refreshToken) throw new AethelgardApiError(401, "missing_refresh_token");
    const bundle = await this.request<AuthBundle>("POST", "/v1/auth/refresh", { refreshToken: this.refreshToken }, false);
    this.setTokens(bundle);
    this.onAuthBundle?.(bundle);
    return bundle;
  }

  async createRoom(region = "asia"): Promise<RoomResponse> {
    return this.roomRequest("POST", "/v1/coop/rooms", { region });
  }

  async joinRoom(roomCode: string): Promise<RoomResponse> {
    return this.roomRequest("POST", `/v1/coop/rooms/${this.normalizeRoomCode(roomCode)}/join`, {});
  }

  async getRoom(roomCode: string): Promise<RoomResponse> {
    return this.roomRequest("GET", `/v1/coop/rooms/${this.normalizeRoomCode(roomCode)}`);
  }

  async heartbeat(roomCode: string, input: HeartbeatInput): Promise<RoomResponse> {
    const payload: HeartbeatInput = {
      playerX: this.clamp(input.playerX, -0.9, 0.9),
      playerY: this.clamp(input.playerY, -0.5, 0.52),
      atTower: Boolean(input.atTower),
      towerRevision: Math.max(0, Math.floor(input.towerRevision)),
    };
    return this.roomRequest("POST", `/v1/coop/rooms/${this.normalizeRoomCode(roomCode)}/heartbeat`, payload);
  }

  async leaveRoom(roomCode: string): Promise<void> {
    await this.request<void>("DELETE", `/v1/coop/rooms/${this.normalizeRoomCode(roomCode)}/leave`);
  }

  async combat(roomCode: string, action: "attack" | "heavy_attack", targetId = "forest_warden", requestId = this.nextRequestId("combat")): Promise<CombatResult> {
    if (!/^[A-Za-z0-9_-]{8,80}$/.test(requestId)) throw new Error("requestId must match [A-Za-z0-9_-]{8,80}");
    return this.request<CombatResult>("POST", `/v1/coop/rooms/${this.normalizeRoomCode(roomCode)}/combat`, { requestId, action, targetId });
  }

  async gather(roomCode: string, resourceId: "forest_cache" | "root_cache" | "warden_stone", requestId = this.nextRequestId("gather")): Promise<InventoryResult> {
    return this.inventory(roomCode, { requestId, operation: "gather", resourceId });
  }

  async craft(roomCode: string, requestId = this.nextRequestId("craft")): Promise<InventoryResult> {
    return this.inventory(roomCode, { requestId, operation: "craft" });
  }

  private async inventory(roomCode: string, body: Record<string, string>): Promise<InventoryResult> {
    if (!/^[A-Za-z0-9_-]{8,80}$/.test(body.requestId)) throw new Error("requestId must match [A-Za-z0-9_-]{8,80}");
    return this.request<InventoryResult>("POST", `/v1/coop/rooms/${this.normalizeRoomCode(roomCode)}/inventory`, body);
  }

  private async roomRequest(method: string, path: string, body?: unknown): Promise<RoomResponse> {
    const response = await this.request<RoomResponse>(method, path, body);
    this.onRoomSnapshot?.(response);
    return response;
  }

  private async request<T>(method: string, path: string, body?: unknown, allowRefresh = true): Promise<T> {
    const headers: Record<string, string> = { Accept: "application/json" };
    if (body !== undefined) headers["Content-Type"] = "application/json";
    if (this.accessToken) headers.Authorization = `Bearer ${this.accessToken}`;

    const response = await this.fetchImpl(`${this.baseUrl}${path}`, {
      method,
      headers,
      body: body === undefined ? undefined : JSON.stringify(body),
    });

    if (response.ok) {
      if (response.status === 204) return undefined as T;
      return (await response.json()) as T;
    }

    const payload = await response.json().catch(() => ({ error: "http_error" }));
    const error = new AethelgardApiError(response.status, String(payload.error ?? "http_error"));
    if (allowRefresh && response.status === 401 && this.refreshToken && path !== "/v1/auth/refresh") {
      await this.refresh();
      return this.request<T>(method, path, body, false);
    }
    throw error;
  }

  private normalizeRoomCode(roomCode: string): string {
    const normalized = roomCode.trim().toUpperCase();
    if (!/^[A-Z0-9]{6}$/.test(normalized)) throw new Error("roomCode must be six uppercase letters or digits");
    return normalized;
  }

  private nextRequestId(prefix: string): string {
    this.requestCounter += 1;
    return `${prefix}-${Date.now()}-${this.requestCounter}`;
  }

  private clamp(value: number, min: number, max: number): number {
    return Math.min(max, Math.max(min, Number.isFinite(value) ? value : min));
  }
}

File: client-sdk/unity/AethelgardCoopClient.cs
using System;
using System.Collections;
using System.Threading.Tasks;
using UnityEngine;
using UnityEngine.Networking;

namespace Aethelgrad.Net
{
    [Serializable]
    public sealed class AuthBundle
    {
        public string accessToken;
        public string refreshToken;
        public string tokenType;
        public string accountId;
        public string accountType;
        public string expiresAt;
        public string refreshExpiresAt;
    }

    [Serializable]
    public sealed class Participant
    {
        public string accountId;
        public float playerX;
        public float playerY;
        public bool atTower;
        public int towerRevision;
    }

    [Serializable]
    public sealed class RoomSnapshot
    {
        public string code;
        public string region;
        public int maxPlayers;
        public float worldTime;
        public int towerRevision;
        public int bossHealth;
        public int combatRevision;
    }

    [Serializable]
    public sealed class RoomResponse
    {
        public RoomSnapshot room;
        public Participant[] participants;
    }

    [Serializable]
    public sealed class CombatResult
    {
        public bool accepted;
        public string action;
        public string targetId;
        public int damage;
        public int bossHealth;
        public int combatRevision;
    }

    [Serializable]
    public sealed class InventoryState
    {
        public int wood;
        public int fiber;
        public int stone;
        public bool emberKit;
    }

    [Serializable]
    public sealed class InventoryResult
    {
        public bool accepted;
        public string operation;
        public InventoryState inventory;
        public int inventoryRevision;
    }

    [Serializable]
    private sealed class ErrorResponse
    {
        public string error;
    }

    [Serializable]
    private sealed class GoogleIdTokenRequest
    {
        public string idToken;
        public GoogleIdTokenRequest(string value) { idToken = value; }
    }

    [Serializable]
    private sealed class GuestKeyRequest
    {
        public string guestKey;
        public GuestKeyRequest(string value) { guestKey = value; }
    }

    [Serializable]
    private sealed class RefreshRequest
    {
        public string refreshToken;
        public RefreshRequest(string value) { refreshToken = value; }
    }

    [Serializable]
    private sealed class RegionRequest
    {
        public string region;
        public RegionRequest(string value) { region = value; }
    }

    [Serializable]
    private sealed class HeartbeatRequest
    {
        public float playerX;
        public float playerY;
        public bool atTower;
        public int towerRevision;
    }

    [Serializable]
    private sealed class CombatRequest
    {
        public string requestId;
        public string action;
        public string targetId;
    }

    [Serializable]
    private sealed class InventoryRequest
    {
        public string requestId;
        public string operation;
        public string resourceId;
    }

    public sealed class AethelgardApiException : Exception
    {
        public long StatusCode { get; private set; }
        public string ErrorCode { get; private set; }

        public AethelgardApiException(long statusCode, string errorCode)
            : base($"AETHELGRAD API {statusCode}: {errorCode}")
        {
            StatusCode = statusCode;
            ErrorCode = errorCode;
        }
    }

    public sealed class AethelgardCoopClient
    {
        private readonly string baseUrl;
        private string accessToken;
        private string refreshToken;
        private int requestCounter;

        public event Action<AuthBundle> AuthBundleReceived;
        public event Action<RoomResponse> RoomSnapshotReceived;

        public AethelgardCoopClient(string serviceBaseUrl, string existingAccessToken = null, string existingRefreshToken = null)
        {
            baseUrl = serviceBaseUrl.TrimEnd('/');
            accessToken = existingAccessToken;
            refreshToken = existingRefreshToken;
        }

        public void SetTokens(AuthBundle bundle)
        {
            accessToken = bundle.accessToken;
            refreshToken = bundle.refreshToken;
        }

        public void ClearTokens()
        {
            accessToken = null;
            refreshToken = null;
        }

        public async Task<AuthBundle> ExchangeGoogleIdTokenAsync(string idToken)
        {
            AuthBundle bundle = await SendJsonAsync<AuthBundle>("POST", "/v1/auth/google-id-token/exchange", new GoogleIdTokenRequest(idToken), false);
            SetTokens(bundle);
            AuthBundleReceived?.Invoke(bundle);
            return bundle;
        }

        public async Task<AuthBundle> AuthenticateGuestAsync(string guestKey)
        {
            if (string.IsNullOrEmpty(guestKey) || !System.Text.RegularExpressions.Regex.IsMatch(guestKey, "^[A-Za-z0-9_-]{32,128}$"))
                throw new ArgumentException("guestKey must be a base64url secret with 32-128 characters");
            AuthBundle bundle = await SendJsonAsync<AuthBundle>("POST", "/v1/auth/guest", new GuestKeyRequest(guestKey), false);
            SetTokens(bundle);
            AuthBundleReceived?.Invoke(bundle);
            return bundle;
        }

        public async Task<AuthBundle> RefreshAsync()
        {
            if (string.IsNullOrEmpty(refreshToken)) throw new AethelgardApiException(401, "missing_refresh_token");
            AuthBundle bundle = await SendJsonAsync<AuthBundle>("POST", "/v1/auth/refresh", new RefreshRequest(refreshToken), false);
            SetTokens(bundle);
            AuthBundleReceived?.Invoke(bundle);
            return bundle;
        }

        public Task<RoomResponse> CreateRoomAsync(string region = "asia")
        {
            return RoomRequestAsync("POST", "/v1/coop/rooms", new RegionRequest(region));
        }

        public Task<RoomResponse> JoinRoomAsync(string roomCode)
        {
            return RoomRequestAsync("POST", "/v1/coop/rooms/" + NormalizeRoomCode(roomCode) + "/join", new object());
        }

        public Task<RoomResponse> GetRoomAsync(string roomCode)
        {
            return RoomRequestAsync("GET", "/v1/coop/rooms/" + NormalizeRoomCode(roomCode), null);
        }

        public Task<RoomResponse> HeartbeatAsync(string roomCode, float playerX, float playerY, bool atTower, int towerRevision)
        {
            HeartbeatRequest payload = new HeartbeatRequest
            {
                playerX = Mathf.Clamp(playerX, -0.90f, 0.90f),
                playerY = Mathf.Clamp(playerY, -0.50f, 0.52f),
                atTower = atTower,
                towerRevision = Mathf.Max(0, towerRevision)
            };
            return RoomRequestAsync("POST", "/v1/coop/rooms/" + NormalizeRoomCode(roomCode) + "/heartbeat", payload);
        }

        public async Task LeaveRoomAsync(string roomCode)
        {
            await SendJsonAsync<object>("DELETE", "/v1/coop/rooms/" + NormalizeRoomCode(roomCode) + "/leave", null);
        }

        public Task<CombatResult> CombatAsync(string roomCode, string action, string targetId = "forest_warden", string requestId = null)
        {
            if (action != "attack" && action != "heavy_attack") throw new ArgumentException("action must be attack or heavy_attack");
            CombatRequest payload = new CombatRequest
            {
                requestId = requestId ?? NextRequestId("combat"),
                action = action,
                targetId = targetId
            };
            return SendJsonAsync<CombatResult>("POST", "/v1/coop/rooms/" + NormalizeRoomCode(roomCode) + "/combat", payload);
        }

        public Task<InventoryResult> GatherAsync(string roomCode, string resourceId, string requestId = null)
        {
            if (resourceId != "forest_cache" && resourceId != "root_cache" && resourceId != "warden_stone") throw new ArgumentException("Unknown resourceId");
            InventoryRequest payload = new InventoryRequest
            {
                requestId = requestId ?? NextRequestId("gather"),
                operation = "gather",
                resourceId = resourceId
            };
            return SendJsonAsync<InventoryResult>("POST", "/v1/coop/rooms/" + NormalizeRoomCode(roomCode) + "/inventory", payload);
        }

        public Task<InventoryResult> CraftAsync(string roomCode, string requestId = null)
        {
            InventoryRequest payload = new InventoryRequest
            {
                requestId = requestId ?? NextRequestId("craft"),
                operation = "craft",
                resourceId = null
            };
            return SendJsonAsync<InventoryResult>("POST", "/v1/coop/rooms/" + NormalizeRoomCode(roomCode) + "/inventory", payload);
        }

        private async Task<RoomResponse> RoomRequestAsync(string method, string path, object body)
        {
            RoomResponse response = await SendJsonAsync<RoomResponse>(method, path, body);
            RoomSnapshotReceived?.Invoke(response);
            return response;
        }

        private async Task<T> SendJsonAsync<T>(string method, string path, object body, bool allowRefresh = true)
        {
            string payload = body == null ? null : JsonUtility.ToJson(body);
            using (UnityWebRequest request = new UnityWebRequest(baseUrl + path, method))
            {
                if (payload != null)
                {
                    byte[] bytes = System.Text.Encoding.UTF8.GetBytes(payload);
                    request.uploadHandler = new UploadHandlerRaw(bytes);
                    request.SetRequestHeader("Content-Type", "application/json");
                }
                request.downloadHandler = new DownloadHandlerBuffer();
                request.SetRequestHeader("Accept", "application/json");
                if (!string.IsNullOrEmpty(accessToken)) request.SetRequestHeader("Authorization", "Bearer " + accessToken);

                UnityWebRequestAsyncOperation operation = request.SendWebRequest();
                while (!operation.isDone) await Task.Yield();

                if (request.result == UnityWebRequest.Result.Success && request.responseCode >= 200 && request.responseCode < 300)
                {
                    if (request.responseCode == 204 || typeof(T) == typeof(object)) return default(T);
                    return JsonUtility.FromJson<T>(request.downloadHandler.text);
                }

                ErrorResponse error = null;
                try { error = JsonUtility.FromJson<ErrorResponse>(request.downloadHandler.text); } catch { }
                string errorCode = string.IsNullOrEmpty(error?.error) ? "http_error" : error.error;
                if (allowRefresh && request.responseCode == 401 && !string.IsNullOrEmpty(refreshToken) && path != "/v1/auth/refresh")
                {
                    await RefreshAsync();
                    return await SendJsonAsync<T>(method, path, body, false);
                }
                throw new AethelgardApiException(request.responseCode, errorCode);
            }
        }

        private string NormalizeRoomCode(string roomCode)
        {
            string normalized = roomCode.Trim().ToUpperInvariant();
            if (normalized.Length != 6) throw new ArgumentException("roomCode must contain six characters");
            for (int i = 0; i < normalized.Length; i++)
            {
                char c = normalized[i];
                if (!(c >= 'A' && c <= 'Z') && !(c >= '0' && c <= '9')) throw new ArgumentException("roomCode must be alphanumeric");
            }
            return normalized;
        }

        private string NextRequestId(string prefix)
        {
            requestCounter++;
            return $"{prefix}-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}-{requestCounter}";
        }
    }
}
